File: android/src/main/kotlin/com/nanofuxion/tamerrouter/TamerRouterNativeModule.kt
package com.nanofuxion.tamerrouter

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.lynx.jsbridge.LynxMethod
import com.lynx.jsbridge.LynxModule
import com.lynx.react.bridge.Callback
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.react.bridge.JavaOnlyMap
import com.lynx.tasm.behavior.LynxContext
import org.json.JSONObject
import org.json.JSONTokener

/**
 * Host wiring: back button callback registration and JS history persistence.
 * JS registers a callback via [registerBackButtonListener] that native invokes when hardware back is pressed.
 * The activity calls [invokeBackButtonCallback] when back is pressed.
 */
class TamerRouterNativeModule(context: Context) : LynxModule(context) {

    init {
        instance = this
        if (context is LynxContext) {
            Companion.lynxContext = context
        }
    }

    companion object {
        private const val TAG = "TamerRouterNative"
        private const val BACK_TIMEOUT_MS = 1000L

        @Volatile
        var instance: TamerRouterNativeModule? = null
            private set

        @Volatile
        private var lynxContext: LynxContext? = null

        @Volatile
        private var historyStateJson: String = """{"entries":["/"],"index":0}"""

        @Volatile
        private var hostSessionId: String = createSessionId()

        @Volatile
        private var historyStateSessionId: String = hostSessionId

        private val mainHandler = Handler(Looper.getMainLooper())

        private fun createSessionId(): String {
            return "${System.currentTimeMillis().toString(36)}-${java.util.UUID.randomUUID().toString().take(8)}"
        }

        /**
         * Attachment point for the LynxView (for future reference or extensions).
         * Not exposed as @LynxMethod to avoid parameter type issues.
         */
        fun attachHostView(view: Any?) {
            // Placeholder for potential future use
        }

        /**
         * Called when hardware back is pressed.
         * Emits a global event and invokes the registered JS callback.
         */
        fun requestBack(fallback: ((Boolean) -> Unit)? = null) {
            val module = instance
            if (module == null) {
                fallback?.invoke(false)
                return
            }
            module.sendBackEvent(fallback)
        }

    }

    @Volatile
    private var backButtonCallback: (Callback)? = null

    @Volatile
    private var pendingBackCallback: ((Boolean) -> Unit)? = null

    private var backTimeoutRunnable: Runnable? = null

    /**
     * Registers a JS callback to be invoked when hardware back is pressed.
     */
    @LynxMethod
    fun registerBackButtonListener(callback: Callback) {
        Log.d(TAG, "Registering back button listener, callback=${if (callback != null) "SET" else "NULL"}")
        backButtonCallback = callback
        Log.d(TAG, "Back button callback stored: ${if (backButtonCallback != null) "SUCCESS" else "FAILED"}")
    }

    /**
     * Called by the activity when hardware back is pressed.
     * Invokes the registered callback if available.
     */
    fun invokeBackCallback() {
        Log.d(TAG, "invokeBackCallback called, backButtonCallback=${if (backButtonCallback != null) "SET" else "NULL"}")
        if (backButtonCallback != null) {
            Log.d(TAG, "Invoking back button callback")
            backButtonCallback?.invoke()
        } else {
            Log.w(TAG, "No back button callback registered!")
        }
    }

    private fun sendBackEvent(fallback: ((Boolean) -> Unit)?) {
        if (pendingBackCallback != null) {
            fallback?.invoke(true)
            return
        }

        pendingBackCallback = fallback

        Log.d(TAG, "requestBack called, lynxContext=${if (lynxContext != null) "SET" else "NULL"}")
        val params = JavaOnlyArray()
        if (lynxContext != null) {
            Log.d(TAG, "Emitting backButtonPressed event")
            lynxContext?.sendGlobalEvent("backButtonPressed", params)
        } else {
            Log.w(TAG, "lynxContext is null, cannot emit event")
        }

        invokeBackCallback()
        scheduleBackTimeout()
    }

    private fun scheduleBackTimeout() {
        backTimeoutRunnable?.let(mainHandler::removeCallbacks)
        val runnable = Runnable {
            val callback = pendingBackCallback
            pendingBackCallback = null
            backTimeoutRunnable = null
            callback?.invoke(false)
        }
        backTimeoutRunnable = runnable
        mainHandler.postDelayed(runnable, BACK_TIMEOUT_MS)
    }

    @LynxMethod
    fun didHandleBack(consumed: Boolean) {
        backTimeoutRunnable?.let(mainHandler::removeCallbacks)
        backTimeoutRunnable = null
        val callback = pendingBackCallback
        pendingBackCallback = null
        callback?.invoke(consumed)
    }

    @LynxMethod
    fun setHistoryState(stateJson: String?) {
        if (!stateJson.isNullOrBlank()) {
            Companion.historyStateJson = stateJson
            Companion.historyStateSessionId = Companion.hostSessionId
        }
    }

    @LynxMethod
    fun consumeHistoryState(callback: Callback) {
        callback.invoke(createHistoryEnvelopeJson())
    }

    private fun createHistoryEnvelopeJson(): String {
        return try {
            val stateValue = JSONTokener(Companion.historyStateJson).nextValue()
            JSONObject()
                .put("hostSessionId", Companion.hostSessionId)
                .put("historySessionId", Companion.historyStateSessionId)
                .put("state", stateValue)
                .toString()
        } catch (_: Exception) {
            JSONObject()
                .put("hostSessionId", Companion.hostSessionId)
                .put("historySessionId", Companion.historyStateSessionId)
                .put("state", JSONObject("""{"entries":["/"],"index":0}"""))
                .toString()
        }
    }
}

File: ios/tamerrouter/tamerrouter/Classes/TamerRouterNativeModule.swift
import Foundation
import UIKit
import Lynx

/// Host wiring only: `tamer-router:back` + `didHandleBack`, and JS history persistence.
/// One back event in flight until JS calls `didHandleBack` (React Native BackHandler–style).
/// Stack visuals are owned by `nav-screen` (`tamer-navigation`), not this module.
@objcMembers
public final class TamerRouterNativeModule: NSObject, LynxModule {

    public static var name: String { "TamerRouterNativeModule" }

    public static var methodLookup: [String: String] {
        [
            "registerBackButtonListener": NSStringFromSelector(#selector(registerBackButtonListener(_:))),
            "setHistoryState": NSStringFromSelector(#selector(setHistoryState(_:))),
            "consumeHistoryState": NSStringFromSelector(#selector(consumeHistoryState(_:))),
            "didHandleBack": NSStringFromSelector(#selector(didHandleBack(_:))),
        ]
    }

    public static weak var instance: TamerRouterNativeModule?

    private static var historyState: String = #"{"entries":["/"],"index":0}"#
    private static var hostSessionId: String = UUID().uuidString
    private static var historyStateSessionId: String = hostSessionId

    @objc public static func attachHostView(_ view: UIView?) {
        if view !== hostView, view != nil {
            hostSessionId = UUID().uuidString
            historyStateSessionId = hostSessionId
            historyState = #"{"entries":["/"],"index":0}"#
        }
        hostView = view
    }

    private static weak var hostView: UIView?

    private weak var lynxContext: LynxContext?
    private var backButtonListener: (() -> Void)?
    private var pendingBackCallback: ((Bool) -> Void)?
    private var backTimeoutWorkItem: DispatchWorkItem?

    public init(param: Any) {
        super.init()
        lynxContext = param as? LynxContext
        Self.instance = self
    }

    public override init() {
        super.init()
        Self.instance = self
    }

    @objc func registerBackButtonListener(_ callback: @escaping () -> Void) {
        backButtonListener = callback
    }

    @objc func setHistoryState(_ stateJson: String) {
        guard !stateJson.isEmpty else { return }
        TamerRouterNativeModule.historyState = stateJson
        TamerRouterNativeModule.historyStateSessionId = TamerRouterNativeModule.hostSessionId
    }

    @objc func consumeHistoryState(_ callback: @escaping (String) -> Void) {
        callback(TamerRouterNativeModule.createHistoryEnvelope())
    }

    @objc func didHandleBack(_ consumed: Bool) {
        backTimeoutWorkItem?.cancel()
        backTimeoutWorkItem = nil
        let cb = pendingBackCallback
        pendingBackCallback = nil
        DispatchQueue.main.async { cb?(consumed) }
    }

    public static func requestBack(callback: @escaping (Bool) -> Void) {
        guard let mod = instance else { callback(false); return }
        mod.sendBackEvent(callback: callback)
    }

    func sendBackEvent(callback: @escaping (Bool) -> Void) {
        guard lynxContext != nil else {
            callback(false)
            return
        }
        if pendingBackCallback != nil {
            callback(true)
            return
        }
        pendingBackCallback = callback
        emitBack()
        scheduleBackTimeout()
    }

    private func scheduleBackTimeout() {
        backTimeoutWorkItem?.cancel()
        let work = DispatchWorkItem { [weak self] in
            guard let self else { return }
            self.backTimeoutWorkItem = nil
            if let cb = self.pendingBackCallback {
                self.pendingBackCallback = nil
                cb(false)
            }
        }
        backTimeoutWorkItem = work
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0, execute: work)
    }

    private static func createHistoryEnvelope() -> String {
        let fallbackState: Any = ["entries": ["/"], "index": 0]
        let parsedState = (try? JSONSerialization.jsonObject(with: Data(historyState.utf8))) ?? fallbackState
        let envelope: [String: Any] = [
            "hostSessionId": hostSessionId,
            "historySessionId": historyStateSessionId,
            "state": parsedState,
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: envelope),
              let json = String(data: data, encoding: .utf8) else {
            return #"{"hostSessionId":"","historySessionId":"","state":{"entries":["/"],"index":0}}"#
        }
        return json
    }

    private func emitBack() {
        if let listener = backButtonListener {
            DispatchQueue.main.async {
                listener()
            }
            return
        }
        guard let ctx = lynxContext else { return }
        emitEvent(ctx, name: "tamer-router:back", payload: "{}")
    }

    private func emitEvent(_ ctx: LynxContext, name: String, payload: String) {
        DispatchQueue.main.async {
            ctx.sendGlobalEvent(name, withParams: [["payload": payload]])
        }
    }
}
